_drafts/zero-to-main-libc.md

```diff
@@ -0,0 +1,385 @@
+a(lib_a-fstatr.o): in function `_fstat_r':
+fstatr.c:(.text._fstat_r+0xe): undefined reference to `_fstat'
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libg_nano.
+a(lib_a-isattyr.o): in function `_isatty_r':
+isattyr.c:(.text._isatty_r+0xc): undefined reference to `_isatty'
+collect2: error: ld returned 1 exit status
+```
+
+Specifically, the compiler is asking for `_fstat`, `_read`, `_lseek`, `_close`, `_write`, and `_sbrk`.
+
+The [newlib documentation](https://sourceware.org/newlib/libc.html#Syscalls) calls these functions “system calls”. In short, they are the handful of things newlib expects the underlying “operating system”. The complete list of them is provided below:
+
+```
+_exit, close, environ, execve, fork, fstat, getpid, isatty, kill, link, lseek, open, read, sbrk, stat, times, unlink, wait, write
+``` 
+
+You’ll notice that several of the syscalls relate to filesystem operation or process control. These do not make much sense in a firmware context, so we’ll often simply provide a stub that returns an error code.
+
+Let’s look at the ones our “Hello, World” example requires. The man pages for them contains all the information we need.
+
+#### fstat
+`fstat`  returns the status of an open file.  The minimal version of this should identify all files as character special devices. This forces one-byte-read at a time.
+```c
+#include <sys/stat.h>
+int fstat(int file, struct stat *st) {
+  st->st_mode = S_IFCHR;
+  return 0;
+}
+```
+
+#### lseek
+
+`lseek` repositions the file offset of the open file associated with the file descriptor `fd` to the argument `offset` according to the directive `whence`. 
+
+Here we can simply return 0, which implies the file is empty.
+
+```c
+int lseek(int file, int offset, int whence) {
+  return 0;
+}
+```
+
+#### close
+
+`close` closes a file descriptor `fd`.
+
+Since no file should have gotten `open`-ed, we can just return an error on close:
+
+```c
+int close(int fd) {
+  return -1;
+}
+```
+
+#### write
+
+This is where things get interesting! `write` writes up to `count` bytes from the buffer starting at `buf` to the file referred to by the file descriptor `fd`.
+
+Functions like `printf` rely on `write` to write bytes to `STDOUT`. In our case, we will want those bytes to be written to serial instead.
+
+On the SAMD21 chip we are using, writing bytes to serial is done using the `usart_serial_putchar` function. We can use it to implement `write`:
+
+```c
+static struct usart_module stdio_uart_module;
+
+int _write (int fd, char *buf, int count) {
```

Comment:
Not necessary but you could also check that fd is one of the [std streams](https://github.com/bminor/newlib/blob/master/newlib/libc/stdio/findfp.c#L90-L118). Should always be 0,1,2 for [posix](https://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html)

Reply:
My initial implementation did that, but it warrants more discussion which isn't all that valuable.